src/runtime/recipe/handle-connection.ts

```diff
@@ -128,20 +133,32 @@ export class HandleConnection {
       }
       return false;
     }
-    if (this.type && this.spec) {
+    if (this.particle.spec && this.name) {
       const connectionSpec = this.spec;
-      if (!connectionSpec.isCompatibleType(this.rawType)) {
+      if (!connectionSpec) {
         if (options && options.errors) {
-          options.errors.set(this, `Type '${this.rawType.toString()} for handle connection '${this.getQualifiedName()}' doesn't match particle spec's type '${connectionSpec.type.toString()}'`);
+          options.errors.set(this, `Connection ${this.name} is not defined by ${this.particle.name}.`);
         }
         return false;
       }
-      if (this.direction !== connectionSpec.direction) {
+      if (this.direction !== connectionSpec.direction &&
+          !(['in', 'out'].includes(this.direction) && connectionSpec.direction === 'inout')) {
         if (options && options.errors) {
           options.errors.set(this, `Direction '${this.direction}' for handle connection '${this.getQualifiedName()}' doesn't match particle spec's direction '${connectionSpec.direction}'`);
         }
         return false;
       }
+      if (!this.rawType) {
+        this._rawType = connectionSpec.type;
+      }
+      if (this.type) {
```

Comment:
this.type will always resolve to this._rawType now, so there's no need for this check as _rawType is assigned directly above. I suspect this could be further simplified too - e.g. no point checking that connectionSpec.type is compatible with connectionSpec.type...

Reply:
updated - only validating if this._resolvedType is set (because it might be different than connectionSpec.type)